packages/core/src/components/TabbedLayout/types.ts

```diff
@@ -14,8 +14,12 @@
  * limitations under the License.
  */
 
+import { TabProps } from '@material-ui/core';
+import * as React from 'react';
+
 export type SubRoute = {
   path: string;
   title: string;
   children: JSX.Element;
+  tabProps?: TabProps<React.ElementType, { component?: React.ElementType }>;
```

Comment:
Thanks for the comments @Rugvip.

Our use case isn't actually the badges themselves for this, that was just something simple to visualize that these changes enable. Our actual use for this is to add more customizability to the whole tab bar:

https://user-images.githubusercontent.com/2392775/108861529-9ad1b400-75ef-11eb-8327-bbc6f11bddef.mp4




I think the largest use case for most users of backstage is to add branding (different color, icons, etc.) to the individual tab and how it's displayed on the entity page tab bar. The current customization possibilities only includes `path` and `title` for these tabs and without exposing the underlying Material UI API the only way forward would be to maintain a parallel class/component/file for at least RoutedTabs, TabbedLayout and HeaderTabs. I think it's good to mention that this change doesn't add any burden to other users and can be left unused if there is no need for modifications. It just exposes the already existing component that is in use for modifications if there is need for those.


We can continue discussion on Discord if you want to bounce around some ideas around this.

Reply:
@Xantier Let's continue here for visibility, but reach out on Discord if you need to share things there.

The visual customizations for branding etc. should be able to be dealt with via the MUI theming, you get access to all kinds of component styles there.

I'm thinking that another approach for this is to completely switch out the entire tabbed page component with a custom one. The possiblity of doing that is actually one of the reasons I'm hesitant to add the props the this component, as any props would have to be re-implemented by the overriding component, and overly broad and unspecific props end up being pretty much impossible to reimplement if they don't fit in.

For now though, what do you think about simply providing a separate and custom table component to use in the app instead of the regular tabbed entity page? No fancy component replacement APIs or anything for now, just simply a component you import from a different package when building the entity page. That component would ofc have to take care to be compatible with the existing tabbed page component though.

Reply:
Thanks @Rugvip 

I think the usage of the tabbed page is a solid design decision since it actually handles so many things nicely already. Also the approach where EntityLayout restricts children to routed components only makes the approach robust for most use cases. There are some issues on composability and possibility for individual modifications on the tabs (this PR) and the kebab-menu on the top right corner. We went this way since it targets the tabs directly and that matches our use case, another option could be to allow the individual larger building blocks (like HeaderTabs.tsx, or EntityContextMenu) to be optionally swapped out to other implementations. That would be a bigger change to the whole, nicely working, rigid EntityLayout approach though.

Switching out the entire tabbed page might get cumbersome in the long run. The addition of Helmet configs, somewhat complex logic around routing and the styling logic based on those would be something that would need to be implemented again on a separate component. That would open up the whole entity page to be recreated the way user wants, but seems to leave a lot of the plumbing to be (re)implemented by the end user on consuming side as well. That is ok for customization side, but leaves the door open for potential bugginess since those would be custom built for each use case. 

I'm not quite certain what components you were thinking about with 'overriding' components, could you elaborate on that? The current implementation proposed here is just drilling through to the underlying Material UI component to add the possibility to use props exposed by the library. I don't think there is need to implement them in any step since the usage is optional. Another approach could be to only expose the `component` prop from Material UI tabs, which makes the element customizable but keeps the API restricted (leaves out `icon`, `disabled` and `classes` props which I feel would be the only ones actually useful.)


I don't have much experience with MUI but I'm not sure branding of individual components based on the runtime presentation of them is something that is supported on the theme layer.